packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -0,0 +1,209 @@
+  }
+
+  /// All possible options that can be searched.
+  ///
+  /// If left null, a custom [search] method must be provided that handles the
+  /// options to be searched on its own.
+  final List<T> options;
+
+  /// The [TextEditingController] that represents the query.
+  final TextEditingController textEditingController;
+
+  /// A search function that takes some [options] and returns a subset as
+  /// results.
+  ///
+  /// If [options] is null, then this field must not be null. This may be the
+  /// case when querying an external service for search results, for example.
+  ///
+  /// Defaults to a simple string-matching search of [options].
+  final AutocompleteSearchFunction<T> search;
+
+  /// The current results being returned by [search].
+  ///
+  /// This is a [ValueNotifier] so that UI may be updated when results change.
+  final ValueNotifier<List<T>> results = ValueNotifier<List<T>>(<T>[]);
+
+  /// Clean up memory created by the AutocompleteController.
+  ///
+  /// Call this when the AutocompleteController is no longer needed, such as in
+  // the dispose method of the widget it was created in.
+  void dispose() {
+    textEditingController.removeListener(_onQueryChanged);
+    // TODO(justinmc): Shouldn't be disposed if it wasn't created here.
+    textEditingController.dispose();
+  }
+
+  // Called when textEditingController reports a change in its value.
+  void _onQueryChanged() {
+    final List<T> resultsValue = search == null
+        ? _searchByString(textEditingController.value.text)
+        : search(textEditingController.value.text);
+    assert(resultsValue != null);
+    results.value = resultsValue;
+  }
+
+  // The default search function, if one wasn't supplied.
+  List<T> _searchByString(String query) {
+    return options
+        .where((T option) => option.toString().contains(query))
+        .toList();
+  }
+}
+
+/// A widget for helping the user to search a list of options and select a
+/// result.
+///
+/// This is a core framework widget with very basic UI. Try using [Autocomplete]
+/// or [AutocompleteCupertino] before resorting to this widget.
+class AutocompleteCore<T> extends StatefulWidget {
+  /// Create an instance of AutocompleteCore.
+  ///
+  /// [autocompleteController], [buildField], and [buildResults] must not be
+  /// null.
+  const AutocompleteCore({
+    @required this.autocompleteController,
+    @required this.buildField,
+    @required this.buildResults,
+  }) : assert(autocompleteController != null),
+       assert(buildField != null),
+       assert(buildResults != null);
+
+  /// The controller that provides access to the main autocomplete state and
+  /// logic.
+  final AutocompleteController<T> autocompleteController;
+
+  /// Builds the field that is used to input the query.
+  final AutocompleteFieldBuilder buildField;
+
+  /// Builds the selectable results of searching.
+  final AutocompleteResultsBuilder<T> buildResults;
```

Comment:
At some point, we may need to provide a very simple example for these two so that users know what they're doing here. Then again, it might not be necessary if we have examples via `Autocomplete` and `AutocompleteCupertino` implementations, as long as they're simple enough.

Reply:
Yeah I agree.  I will add a top level example to AutocompleteCore for this PR at least.  It might be nice to add them to buildField and buildResults as well.

Reply:
I've added an example to AutocompleteCore and AutocompleteController, but I didn't add them to the builders since it's covered by the AutocompleteCore example.  When we implement the Material Autocomplete we should add an example to buildField and buildResults there, because there they have default values and can be used independently of each other.